### DISCUSSION TITLE
Inheritance - Deserialize a child from a parent representation

### DISCUSSION THREAD

**Discussion by OpenGuidou** [FasterXML/jackson-databind#4054] [Q&A]
Hello,

I'm working with the following case:

```
@JsonTypeInfo(defaultImpl = Animal.class, use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type", visible = true)
@JsonSubTypes({
        @JsonSubTypes.Type(value = Animal.class, name = "animal"),
        @JsonSubTypes.Type(value = Dog.class, name = "dog")
})
public class Animal {
    
    private String type;

    private String name;

    @JsonIgnore
    @JsonProperty("type")
    public String getType() {
        return type;
    }

    @JsonProperty("name")
    public String getName() {
        return name;
    }
}

```

```
public class Dog extends Animal {

    private boolean bark;

    @JsonProperty("bark")
    public boolean isBarking() {
        return bark;
    }
}

```

Given the following JSON:
```
{
  "name": "myAnimal",
  "type": "animal"
}
```

I know it's a bit of a stretch use-case, but I would like to work at Dog level in the application reading this JSON.
Once deserialized, as the object built is an Animal, it can no longer be instantiated or casted as something else.
I'm wondering if it would be possible to deserialize it directly as a Dog, and not an Animal.

As the "type" = "animal", when trying to do so I'm getting:

> Could not resolve type id 'animal' as a subtype of `Dog`: Class `Animal` not subtype of `Dog`

And I can't change the "type" to be something else (it's generated by someone else).

Would there be a way to achieve this ?

**Comment by JooHyukKim** [comment]
There may be a solution to your problem. So what you want is to deserialize a JSON object as a Dog object even though the JSON's type field indicates it is an Animal? (with `"type" : "animal")

There is a new configuration feauture in Jackson 2.15, `MapperFeature.REQUIRE_TYPE_ID_FOR_SUBTYPES`. It allows deserialization into concrete subtypes.

**Reply by JooHyukKim** [reply]
> And I can't change the "type" to be something else (it's generated by someone else).

Downside is that above mentioned feature seems only valid if `"type"` field is excluded from the JSON data itself.

```java
    public void testff() throws Exception {
        String data = a2q("{'name': 'myAnimal'}");
        Dog d = JsonMapper.builder()
                .disable(MapperFeature.REQUIRE_TYPE_ID_FOR_SUBTYPES)
                .build()
                .readValue(data, Dog.class);
        
    }
```

**Comment by JooHyukKim** [comment]
May I ask for your opinion on the following, @cowtowncoder?

If we:

1. Disable `MapperFeature.REQUIRE_TYPE_ID_FOR_SUBTYPES`, and
2. Are deserializing into a concrete type (in this case `Dog.class`),

would it make more sense to throw `UnrecognizedPropertyException` on JSON field `"type"`, instead of `InvalidTypeIdException`? What "possibly" makes current implementation counter-intuitive is that someone would think Polymorphic type handling would be hidden away from below deserialization handling. 

Here's an example with the above-mentioned classes:

```java
public void testConcrete() throws Exception {
    String data = a2q("{'name': 'myAnimal','type': 'animal'}");
    Dog d = JsonMapper.builder()
        .disable(MapperFeature.REQUIRE_TYPE_ID_FOR_SUBTYPES)
        .build()
        .readValue(data, Dog.class);
}

**Comment by cowtowncoder** [comment]
As to the original question: no, if type id given is `animal` and it matches `Animal` (but not `Dog`), I do not think it should work in any different way than binding to `Animal`.

`MapperFeature.REQUIRE_TYPE_ID_FOR_SUBTYPES` is only for cases where type id is missing; it does not disable Type Id handling; but in this case it exists (`@JsonIgnore` does not apply to type ids).

I am not dead set against improvements to error messages tho if someone wants to submit a PR. But I don't think handling itself should be changed or changeable in this case.
